### PR TITLE
Expose residuals in WavelengthCalibration1D

### PR DESCRIPTION
**Pull request by cshanahan1**
Add ``fit_residuals`` attribute to ``WavelengthCalibration1D``. These are computed by evaluating the model after the fit is performed, which is done when ``apply_to_spectrum`` is called.

(Closes JDAT-3405)

### CODE REVIEW

File: specreduce/wavelength_calibration.py
Comment (kecnry) [inline specreduce/wavelength_calibration.py:196]: should `_fit_resids` be initialized to `None` in the `__init__` to avoid an error when accessing this before calling `apply_to_spectrum`?  Should that case raise a useful error stating that `apply_to_spectrum` needs to be called first?

File: specreduce/wavelength_calibration.py
Comment (kecnry) [inline specreduce/wavelength_calibration.py:217]: I wonder if this should just be its own cached property so that it knows to clear when the model is manually changed by the user and to save a (small) bit of time when applying to a spectrum but not requesting residuals?

File: specreduce/wavelength_calibration.py
Comment (kecnry) [inline specreduce/wavelength_calibration.py:195]: probably should have a docstring to explain what this is and that its only available after calling `apply_to_spectrum`.

File: specreduce/wavelength_calibration.py
Comment (cshanahan1) [reply specreduce/wavelength_calibration.py:196]: My thought was that if it was not initialized, it WOULD raise an error if apply_to_spectrum wasn't run first.

I could do it this way - initialize it to None, have a check in the getter that if fit_residuals is None, raise a more informative error. Do you prefer that approach?

Comment (kecnry) [reply specreduce/wavelength_calibration.py:196]: Yes, but the error will seem cryptic to the user (just something like "WavelengthCalibration object has no attribute _fit_resids").  I think either what you suggest or the cached_property could both catch the error and suggest to the user what needs to be done from their end.

File: specreduce/wavelength_calibration.py
Comment (cshanahan1) [reply specreduce/wavelength_calibration.py:196]: done

File: specreduce/wavelength_calibration.py
Comment (cshanahan1) [reply specreduce/wavelength_calibration.py]: done

File: specreduce/wavelength_calibration.py
Comment (cshanahan1) [reply specreduce/wavelength_calibration.py:196]: oh actually, it would be accessible after calling either .wcs or .apply_to_spectrum. the fit is done in .wcs.

File: specreduce/wavelength_calibration.py
Comment (cshanahan1) [reply specreduce/wavelength_calibration.py:214]: i also added a test to make sure this error is raised when this happens

Comment (kecnry) [reply specreduce/wavelength_calibration.py:215]: would checking for `self._model is None` for the error get around the need to set `_fit_resids` to False?

File: specreduce/tests/test_wavelength_calibration.py
Comment (rosteen) [inline specreduce/tests/test_wavelength_calibration.py:112]: I think we generally use `match="some text"` in the `pytest.raises` call to make sure that the specific error being raised is the one we expect, in case some other error starts happening at some point.

**Review by rosteen — COMMENTED** [review]
One minor suggestion to use `match` in the error test, other than that it looks good. I don't love the `None` vs `False` distinction to check to see if the residuals are available, but I also don't know that there's a better way to do it...can't check to see if `wcs` is `None` because that would just force the model to fit 😅 .

File: specreduce/wavelength_calibration.py
Comment (cshanahan1) [reply specreduce/wavelength_calibration.py:215]: no, because both ._model and .model will always exist (default is a linear model), just not fit if .wcs isn't accessed.

File: specreduce/tests/test_wavelength_calibration.py
Comment (cshanahan1) [reply specreduce/tests/test_wavelength_calibration.py]: done!

**Review by rosteen — APPROVED** [review]
LGTM, thanks!

**Comment by cshanahan1** [comment]
and yeah, regarding comments on how to raise an error when fit_residuals is accessed with .wcs never being run - we can't check if _fit_resids is None (which is what its initialized to) because that would also fail upon the first time computing residuals. there needs to be some switch flipped by .wcs to indicate that residuals can now be fit, so i can either do what i did or make a new private attribute

**Review by kecnry — APPROVED** [review]
As discussed offline, avoiding the `_fit_resids = False` and needing to call `wcs` or `apply_to_spectrum` _before_ `fit_residuals` will likely result in a slight change in behavior to how successful calls will act, so I'll open that as a follow-up PR for discussion after this is merged (since this accomplishes the immediate goals).  Thanks @cshanahan1!

**Comment by cshanahan1** [comment]
im gonna merge this now since it has two approvals. thanks!

**Pull request merged**
